flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistry.java

```diff
@@ -51,9 +58,9 @@ public void prune(long checkpointID) {}
      * are unregistered when the checkpoint completes, because only one checkpoint is kept for local
      * recovery.
      *
-     * @param upTo lowest CheckpointID which is still valid.
+     * @param latestSubsumedId latest subsumed checkpointId.
      */
-    void discardUpToCheckpoint(long upTo);
+    void discardUpToCheckpoint(long latestSubsumedId);
```

Comment:
Previously, `discardUpToCheckpoint` was called in both `confirm()` and `subsume()`, and in fact this is a duplicate, now it is only called in `subsume()`, and the logic in `discardUpToCheckpoint` also changed (`<` -> `<=`).
If `state.checkpoints.num-retained >1`, this will retain more local files that are not needed, because local recovery only needs one previous chk. 

I fixed this by remove `subsume` and only call `discardUpToCheckpoint` when `confirm()`.